ios/Exponent/Versioned/Core/UniversalModules/EXNotifications/EXScopedNotificationsUtils.m

```diff
@@ -18,4 +18,71 @@ + (BOOL)shouldNotification:(UNNotification *)notification beHandledByExperience:
   return [EXScopedNotificationsUtils shouldNotificationRequest:notification.request beHandledByExperience:experienceId];
 }
 
++ (NSString *)scopedCategoryIdentifierWithId:(NSString *)categoryId forExperience:(NSString *)experienceId
+{
+  NSString *scope = [EXScopedNotificationsUtils escapedString:experienceId];
+  NSString *escapedCategoryId = [EXScopedNotificationsUtils escapedString:categoryId];
+  return [NSString stringWithFormat:@"%@/%@", scope, escapedCategoryId];
+}
+
++ (BOOL)isCategoryId:(NSString *)identifier scopedByExperience:(NSString *)experienceId
+{
+  NSString *scopeFromCategoryId = [EXScopedNotificationsUtils getScopeAndIdentifierFromScopedIdentifier:identifier].scopeKey;
+  return [scopeFromCategoryId isEqualToString:experienceId];
+}
+
++ (ScopedCategoryIdentifierComponents)getScopeAndIdentifierFromScopedIdentifier:(NSString *)scopedIdentifier
+{
+  NSString *scope = @"";
+  NSString *identifier = @"";
+  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^(.*(?<!\\\\)|.*(?:\\\\\\\\)+)/(.*)$"
```

Comment:
needed to add the second case : `.*(?:\\\\\\\\)+` for cases where the scope _ends_ with `\`, and thus the full scoped ID would be
```
experienceId\\/mycategoryname
```

Reply:
I see what you're doing. Here's a way that might be simpler to think about and easier to prove correct:

A valid escaped string consists of:
* unescaped characters, namely everything except for `/` and `\`
* escaped characters, namely `\/` and `\\`.
This regex matches every valid escaped string:
```
(?: # need a group but don't need to capture
  [^/\\] |  # an unescaped character, or
  \\[/\\]   # an escaped character
)*  # as many of these characters as we can read
```

If this regex is run against `@hello\\/world\/`, it will find two matches (`@hello\\` and `world\/`) even though we haven't even told the regex to look for the delimiter in the middle!

In our case, we know that all valid scoped identifiers comprise exactly two escaped strings with a delimiter in the middle, so we conceptually want to write: `^(:escaped_string)/(:escaped_string)$`. Substituting in the regex above, we get:

```
^((?:[^/\\]|\\[/\\])*)/((?:[^/\\]|\\[/\\])*)$
```

Now escaping this for Objective-C string literals, we get:
```objc
@"^((?:[^/\\\\]|\\\\[/\\\\])*)/((?:[^/\\\\]|\\\\[/\\\\])*)$"
```
or
```objc
NSString *pattern = @"^"
                     "((?:[^/\\\\]|\\\\[/\\\\])*)" // escaped scope key
                     "/"                           // delimiter
                     "((?:[^/\\\\]|\\\\[/\\\\])*)" // escaped original category identifier
                     "$";
```

There's a lot of escaping going on, but as you pop off each layer of escaping you get to a regex that is composed of pieces that are easier to understand. It also runs very efficiently with no backtracking at all.

```suggestion
  NSString *pattern = @"^"
                       "((?:[^/\\\\]|\\\\[/\\\\])*)" // escaped scope key
                       "/"                           // delimiter
                       "((?:[^/\\\\]|\\\\[/\\\\])*)" // escaped original category identifier
                       "$";
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:pattern
```

Reply:
very nice!

re-tested with this change and all works as expected